dozer-orchestrator/src/pipeline/sinks.rs

```diff
@@ -364,54 +366,107 @@ impl Sink for CacheSink {
-                    ExecutionError::SinkError(SinkError::CacheDeleteFailed(
-                        endpoint_name.clone(),
-                        Box::new(e),
-                    ))
-                })?;
-                old.version = Some(version);
-
-                if let Some(notifier) = &self.notifier {
-                    let op =
-                        types_helper::map_delete_operation(self.api_endpoint.name.clone(), old);
-                    try_send(&notifier.1, op)?;
+                let result = self.cache.delete(&key);
+                match result {
+                    Ok(version) => {
+                        old.version = Some(version);
+
+                        if let Some(notifier) = &self.notifier {
+                            let op = types_helper::map_delete_operation(
+                                self.api_endpoint.name.clone(),
+                                old,
+                            );
+                            try_send(&notifier.1, op)?;
+                        }
+                    }
+                    Err(e) => {
+                        ConflictResolver::resolve_delete_error(
+                            old,
+                            schema,
+                            e,
+                            self.api_endpoint.conflict_resolution.clone(),
+                        )
+                        .map_err(|e| {
+                            ExecutionError::SinkError(SinkError::CacheDeleteFailed(
+                                endpoint_name.clone(),
+                                Box::new(e),
+                            ))
+                        })?;
+                    }
                 }
             }
             Operation::Insert { mut new } => {
                 new.schema_id = schema.identifier;
-                let id = self.cache.insert(&mut new).map_err(|e| {
-                    ExecutionError::SinkError(SinkError::CacheInsertFailed(
-                        endpoint_name.clone(),
-                        Box::new(e),
-                    ))
-                })?;
-
-                if let Some(notifier) = &self.notifier {
-                    let op =
-                        types_helper::map_insert_operation(self.api_endpoint.name.clone(), new, id);
-                    try_send(&notifier.1, op)?;
+                let result = self.cache.insert(&mut new);
+
+                match result {
+                    Ok(id) => {
+                        if let Some(notifier) = &self.notifier {
+                            let op = types_helper::map_insert_operation(
+                                self.api_endpoint.name.clone(),
+                                new,
+                                id,
+                            );
+                            try_send(&notifier.1, op)?;
+                        }
+                    }
+                    Err(e) => {
+                        let new_op = ConflictResolver::resolve_insert_error(
```

Comment:
This is not a proper `upsert` because after `insert` failed and before processing the newly generated `update`, the record may have been deleted by someone else.